File: typescript/patterns-use-cases/src/sagas/activities/payment_client.ts
import {randomUUID} from "node:crypto";
import {TerminalError} from "@restatedev/restate-sdk";

export const paymentClient = {
  charge: async (request: { paymentInfo: { cardNumber: string, amount: number }, paymentId: string }) => {
    if (Math.random() < 0.5) {
      console.error("This payment should never be accepted! Aborting booking.");
      throw new TerminalError("This payment could not be accepted!");
    }
    if (Math.random() < 0.8) {
      console.error("A payment failure happened! Will retry...");
      throw new Error("A payment failure happened! Will retry...");
    }
    console.info(`Payment ${request.paymentId} processed`);
  },

  refund: async (req: { paymentId: string }) => {
    console.info(`Payment ${req.paymentId} refunded`);
  },
};

File: go/patterns-use-cases/src/sagas/activities/paymentclient.go
package activities

import (
	"errors"
	"fmt"
	restate "github.com/restatedev/sdk-go"
	"log/slog"
	"math/rand"
)

type PaymentInfo struct {
	CardNumber string  `json:"card_number"`
	Amount     float64 `json:"amount"`
}

type PaymentClient struct{}

func (PaymentClient) Charge(paymentId string, _paymentRequest PaymentInfo) error {
	if rand.Float64() < 0.5 {
		slog.Error("👻 This payment should never be accepted! Aborting booking.")
		return restate.TerminalError(errors.New("👻 this payment could not be accepted"))
	}
	if rand.Float64() < 0.8 {
		slog.Error("👻 A payment failure happened! Will retry...")
		return errors.New("👻 a payment failure happened! Will retry")
	}
	slog.Info(fmt.Sprintf("👻 Payment %s processed", paymentId))
	return nil
}

func (PaymentClient) Refund(paymentId string) error {
	slog.Info(`Refunded payment: ` + paymentId)
	return nil
}
